File: 2022-07-26/incr_subseq_with_drops/sol/sol.py
#!/usr/bin/env python3
import sys

MOD=1000000007

g,N,K = list(map(int,input().strip().split()))
panino = list(map(int,input().strip().split()))
P= [ [0] * N for k in range(K+1) ]  #  P[i][k] = massima lunghezza di una sottosequenza crescente con ESATTAMENTE k crolli che inizi prendendo l'elemento i-esimo come suo primo elemento.
num_opts= [ [None] * N for k in range(K+1) ] #  num_opts[i][k] = numero di soluzioni ottime al probema P[i][k]. Ossia numero di sottosequenza crescente con al più k crolli di massima lunghezza tra quelle che iniziano prendendo l'elemento i-esimo come loro primo elemento.
opt_val=0
num_global_opts=1 # vi è una solo sottoinsieme vuoto dell'insieme degli indici i in range(N)
for i in range(N-1,-1,-1):
  P[0][i] = 1    # soluzione: prendere solo quel panino
  num_opts[0][i] = 1 # vi è un solo modo di prendere solo quel panino
  for k in range(K+1):
      for j in range(i+1,N):   
        if panino[j]>panino[i]: # il panino j candida per essere il secondo panino
            if 1+P[k][j] > max(k,P[k][i]):
                P[k][i] = 1+P[k][j]
                num_opts[k][i] = num_opts[k][j]
                #print(f"reset j={j},[k,i]=[{k},{i}],num_opts[k][i]={num_opts[k][i]}")
            elif 1+P[k][j] == P[k][i] > k:
                num_opts[k][i] =  (num_opts[k][i] + num_opts[k][j]) % MOD
                #print(f"added j={j},[k,i]=[{k},{i}],num_opts[k][i]={num_opts[k][i]}")
        elif k>0:
            if 1+P[k-1][j] > max(k,P[k][i]):
                P[k][i] = 1+P[k-1][j]
                num_opts[k][i] = num_opts[k-1][j]
                #print(f"reset j={j},[k,i]=[{k},{i}],num_opts[k][i]={num_opts[k][i]}")
            elif 1+P[k-1][j] == P[k][i]  > k:
                num_opts[k][i] = (num_opts[k][i] + num_opts[k-1][j]) % MOD
                #print(f"added j={j},[k,i]=[{k},{i}],num_opts[k][i]={num_opts[k][i]}")
      if P[k][i] > opt_val:
          opt_val = P[k][i]
          num_global_opts = num_opts[k][i]
      elif P[k][i] == opt_val:
          num_global_opts = (num_global_opts + num_opts[k][i]) % MOD

#print(f"input    ={panino}")
#for k in range(K+1):
#  print(f"P[{k}]       ={P[k]}")
#for k in range(K+1):
#  print(f"num_opts[{k}]={num_opts[k]}")
          
if g==1:
    print(opt_val)
else:    
    print(num_global_opts)
exit(0)

File: 2022-07-26/incr_subseq_with_drops/sol/sol.cpp
/**
 *  File: sol.cpp
 *  Soluzione per incr_subseq_with_drops, quadratica nel tempo e lineare nello spazio
 *  Romeo Rizzi, 2022-07-23
 *
 */

#include <cassert>
#include <cstdio>

const int MAXN = 200;
const int MAXK = 200;
const int MOD=1000000007;

typedef long long unsigned int llu;

int g,N,K;
int s[MAXN]; // Gli indici della sequenza in input partono da 0.
int max_from[MAXK+1][MAXN+1]; // max_from[k][i] = massima lunghezza di una sottosequenza di al più k drops che abbia s[i] come suo primo elemento (se i=n la lunghezza è 0)
int num_opts_from[MAXK+1][MAXN+1]; // num_opts_from[k][i] = numero di sottosequenze di al più k drops che abbiano s[i] come loro primo elemento aventi massima lunghezza (se i=n tale numero è 1 e corrisponde alla sequenza vuota)

int main() {
  scanf("%d %d %d", &g, &N, &K);
  for(int i = 0; i < N; i++)
    scanf("%d", &s[i]);
  for(int k=0; k<=MAXK; k++) num_opts_from[k][N] = 1;
  for(int i=N-1; i >= 0; i--) {
    for(int k=0; k<=MAXK; k++) {
      max_from[k][i] = 0;
      num_opts_from[k][i]=1;
      for(int j=N-1; j>i; j--) {
        if(s[j]>s[i]) { //s_j candida per essere il prossimo elemento della sottosequenza senza incremento in drops
          if(1+max_from[k][j]>max_from[k][i]) {
  	      max_from[k][i] = 1+max_from[k][j];
              num_opts_from[k][i] = num_opts_from[k][j];
	  }
          else if(1+max_from[k][j]==max_from[k][i]) {
	    num_opts_from[k][i] = (num_opts_from[k][i] + num_opts_from[k][j]) % MOD;
	  }
	}
	else if(k>0) {
          if(1+max_from[k-1][j]>max_from[k][i]) {
  	      max_from[k][i] = 1+max_from[k-1][j];
              num_opts_from[k][i] = num_opts_from[k-1][j];
	  }
          else if(1+max_from[k-1][j]==max_from[k][i]) {
              num_opts_from[k][i] = (num_opts_from[k][i] + num_opts_from[k-1][j]) % MOD;
	  }
	}
      }
    }
  }
  if(g==1)
    printf("%d\n", max_from[K][0]);
  else if(g==2)
    printf("%d\n", num_opts_from[K][0]);
  else
    assert(0);
        
  return 0;
}

